mam/src/mam.rs

```diff
@@ -39,116 +37,148 @@ where
-        curl1.reset();
-        merkle::root(&next_addrs[0], &siblings, 0, curl1)
-    };
-
-    curl1.reset();
-    curl2.reset();
-
-    merkle::key(seed, start + index, security, &mut key, curl1);
-    curl1.reset();
-    let mut payload = sign::<C, CB, H>(message, &next, &key, &siblings, security, curl1, bcurl);
-
-    {
-        // generate the message key for a root and an index.
-        let mut channel_key: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
-        // It copies `root` to `channel_key`
-        channel_key.clone_from_slice(&root[..HASH_LENGTH]);
-        // add `index` to it.
-        add_assign(&mut channel_key, index as isize);
-        mask::<C>(&mut payload, &channel_key, curl2);
+        curl_bak.clone_from_slice(&sponge.state());
+        sponge.reset();
+        iss::signature(&curl_bak[..HASH_LENGTH], &mut key, sponge);
+        sponge.state_mut().clone_from_slice(&curl_bak);
+        payload.extend_from_slice(&key);
+        payload.extend_from_slice(&pascal::encode(siblings.len() / HASH_LENGTH));
+        payload.extend_from_slice(&siblings);
+        let end = payload.len();
+        mask_slice(&mut payload[cursor..end], sponge);
+        sponge.reset();
     }
-    (payload, root)
+    payload
+}
+enum UnmaskPascalStatus {
+    /// Message Hash did not have any hamming weight of zero
+    Searching,
+    /// Signature did not match expected root
+    Finalizing,
+    /// Array was too short
+    Done,
 }
 
 pub fn parse<C>(
     payload: &[Trit],
+    side_key: &[Trit],
     root: &[Trit],
     index: usize,
-    curl1: &mut C,
-    curl2: &mut C,
+    sponge: &mut C,
 ) -> Result<(Vec<Trit>, Vec<Trit>), MamError>
 where
     C: Curl<Trit>,
 {
-    let mut channel_key: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
-    channel_key.clone_from_slice(&root[..HASH_LENGTH]);
-    add_assign(&mut channel_key, index as isize);
-    let mut unmasked_payload = payload.to_vec();
-    unmask::<C>(&mut unmasked_payload, &channel_key, curl1);
-
-    curl1.reset();
-    authenticate::<C>(&unmasked_payload, root, index, curl1, curl2)
+    let mut hash: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
+    sponge.reset();
+    message_key(side_key, root, index, sponge);
+
+    let (message_length, message_length_end) = pascal::decode(&payload);
+    if message_length > payload.len() {
+        return Err(MamError::ArrayOutOfBounds);
+    }
+    sponge.absorb(&payload[..message_length_end]);
+    let mut out: Vec<Trit> = (&payload[message_length_end..payload.len()])
+        .clone()
+        .to_vec();
+    let mut pos = {
+        unmask_slice(&mut out[..message_length], sponge);
+        message_length
+    };
+    unmask_slice(&mut out[pos..pos + HASH_LENGTH / 3], sponge);
+    pos += HASH_LENGTH / 3;
+    let mut hmac: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
+    hmac.clone_from_slice(&sponge.rate());
+    let security = iss::checksum_security(&hmac);
+    let out_len = out.len();
+    unmask_slice(&mut out[pos..out_len], sponge);
```

Comment:
You don't need an explicit `out_len` here, just do `out[pos..]`.
That way, the variable becomes obsolete as well.